mapie/regression/estimator.py

```diff
@@ -0,0 +1,504 @@
+
+    estimators_: list
+        List of out-of-folds estimators.
+
+    k_: ArrayLike
+        - Array of nans, of shape (len(y), 1) if ``cv`` is ``"prefit"``
+          (defined but not used)
+        - Dummy array of folds containing each training sample, otherwise.
+          Of shape (n_samples_train, cv.get_n_splits(X_train, y_train)).
+    """
+    no_agg_cv_ = ["prefit", "split"]
+    no_agg_methods_ = ["naive", "base"]
+    fit_attributes = [
+        "single_estimator_",
+        "estimators_",
+        "k_",
+    ]
+
+    def __init__(
+        self,
+        estimator: Optional[RegressorMixin],
+        method: str,
+        cv: Optional[Union[int, str, BaseCrossValidator]],
+        agg_function: Optional[str],
+        n_jobs: Optional[int],
+        random_state: Optional[Union[int, np.random.RandomState]],
+        test_size: Optional[Union[int, float]],
+        verbose: int
+    ):
+        self.estimator = estimator
+        self.method = method
+        self.cv = cv
+        self.agg_function = agg_function
+        self.n_jobs = n_jobs
+        self.random_state = random_state
+        self.test_size = test_size
+        self.verbose = verbose
+
+    @staticmethod
+    def _fit_oof_estimator(
+        estimator: RegressorMixin,
+        X: ArrayLike,
+        y: ArrayLike,
+        train_index: ArrayLike,
+        sample_weight: Optional[ArrayLike] = None,
+    ) -> RegressorMixin:
+        """
+        Fit a single out-of-fold model on a given training set.
+
+        Parameters
+        ----------
+        estimator: RegressorMixin
+            Estimator to train.
+
+        X: ArrayLike of shape (n_samples, n_features)
+            Input data.
+
+        y: ArrayLike of shape (n_samples,)
+            Input labels.
+
+        train_index: ArrayLike of shape (n_samples_train)
+            Training data indices.
+
+        sample_weight: Optional[ArrayLike] of shape (n_samples,)
+            Sample weights. If None, then samples are equally weighted.
+            By default ``None``.
+
+        Returns
+        -------
+        Tuple[RegressorMixin, NDArray, ArrayLike]
+
+        - [0]: RegressorMixin, fitted estimator
+        - [1]: NDArray of shape (n_samples_val,),
+          estimator predictions on the validation fold.
+        - [2]: ArrayLike of shape (n_samples_val,),
+          validation data indices.
+        """
+        X_train = _safe_indexing(X, train_index)
+        y_train = _safe_indexing(y, train_index)
+        if not (sample_weight is None):
+            sample_weight = _safe_indexing(sample_weight, train_index)
+            sample_weight = cast(NDArray, sample_weight)
+
+        estimator = fit_estimator(
+            estimator, X_train, y_train, sample_weight=sample_weight
+        )
+        return estimator
+
+    def fit(
+        self,
+        X: ArrayLike,
+        y: ArrayLike,
+        sample_weight: Optional[ArrayLike] = None,
+    ) -> EnsembleRegressor:
```

Comment:
UPD: the `fit` method must be at the end of the file, just before the `predict` method (as done in `MapieRegressor`). Better for reading.

Reply:
Done